enhancements/network/external-lb-vips.md

```diff
@@ -0,0 +1,213 @@
+
+### User Stories
+
+As a deployer of a IPI OpenShift cluster, I want to place my control plane nodes and compute nodes on 3 or
+more subnets at installation time. Therefore I need to be able to configure an external load balancer for
+the VIPs.
+
+### Goals
+
+Allow to replace the internal loadbalancer by an external one and support this scenario when the customer
+know what they're doing.
+
+### Non-Goals
+
+* Making this scenario by default at some point. The internal loadbalancer for on-premise platforms will
+  remain the HAproxy/Keepalived/CoreDNS stack.
+* Supporting a mix of both (internal + external loadbalancers for the API & Ingress VIPs), customers
+  will have to choose a scenario.
+
+## Proposal
+
+We will make some change in `install-config.yaml` so customer can configure the
+deployment to use an external loadbalancer if they want.
+Then the components which interact with networking will have to check this
+new field and handle this scenario appropriately.
+
+### Workflow Description
+
+1. The customer will configure their external loadbalancer(s), which includes configuring the VIPs
+   for API and Ingress.
+2. The customer will populate their external DNS server with required records for api, api-int, and ingress.
+3. The customer will write the `install-config.yaml` file and set the new `loadBalancerType` field to `external`, and
+   will not set the fields for apiVIP and ingressVIP. We should fail if they set these mutually exclusive parameters.
+4. In the installer, if the platform does support external loadbalancer, some validation will be run
+   to check that the VIPs can already be resolved (the validation already exists for vsphere, we need to
+    make it re-usable). A new field in `PlatformStatus` will be set.
+5. In machine-config-operator, if LB type is external, the internal loadbalancer will not be
+   deployed.
+6. When using an external loadbalancer, the VIPs fields won't be set in `PlatformStatus` and the
+   OpenShift components won't use it when they're not present and LB type is external.
+7. After the deployment, the IP address of a service (e.g. API) can change, as long as the external
+   loadbalancer is updated and the DNS record has been updated and the results are propagated as well.
+
+### API Extensions
+
+A new field will be added to the `PlatformStatus` of the infrastructure object to tell
+the kind of the loadbalancer.
+
+```golang
+// LoadBalancerType can be either external or internal 
+type LoadBalancerType string
+const (
+	// LoadBalancerTypeExternal
+	LoadBalancerTypeExternal LoadBalancerType = "external"
+	// LoadBalancerTypeInternal
+	LoadBalancerTypeInternal LoadBalancerType = "internal"
+)
+// VSpherePlatformStatus holds the current status of the vSphere infrastructure provider.
+type VSpherePlatformStatus struct {
```

Comment:
Adding this to the platform provider means adding it individually for every provider. However, although support per-provider is required, there is nothing here which prevents it from being implemented for any provider. The principal would be the same in all cases:

* Deploy without a loadbalancer
* User configures the loadbalancer

Additionally, you have specified the status here. Did you intend the spec? Unless there is some unusual convention for the platform provider, status is not intended to be user-editable. It typically contains observed state rather than intended state.

Reply:
Or is this intended to describe only the reported status? If so, you also need to define the spec. When would the status be set, and by what?

Reply:
I actually followed the other PlaformStatus fields for the VIPs, which are set by the installer.
I don't have much opinion nor knowledge on where would be a better place for `LoadBalancerType` to live.

@cybertron @JoelSpeed what do you think?

Reply:
I believe the reason all of these are platform-specific is that they only apply to the on-prem platforms. We don't want to add a LoadBalancerType field that will be present on platforms that don't support it. It does result in a lot of duplication, but unless we're going to add an on-prem section (which we did in MCO, so there is precedent) it's hard to dedupe without things leaking into other platforms.

Everything is in PlatformStatus because up to now we haven't allowed changing any of these values. I think that will need to change going forward, but I don't think LoadBalancerType needs to. I don't feel the need to allow switching between internal and external LBs, and we can always revisit that if there's demand.

Reply:
> Everything is in PlatformStatus because up to now we haven't allowed changing any of these values. I think that will need to change going forward, but I don't think LoadBalancerType needs to. I don't feel the need to allow switching between internal and external LBs, and we can always revisit that if there's demand.

As a counter point, putting it in the status makes it hard to evolve later as we would have to add a spec field and work out what to do about the status field. If we add a spec field now and make it immutable, we could relax that validation later if we desired to do so. To relax a validation you need to make sure all consumers of the API are going to be updated to respect the more relaxed validation, fortunately, we control the consumers of the infrastructure object so we can make sure to do that. Something to consider at least 🤔

Reply:
For now, we agreed to move it to `PlatformSpec` as immutable but if at some point we want to be able to change it's value, it'll be easier to do it if it's already in the Spec and not in Status.